### ISSUE TITLE
Revoking exit node internet access in ACL for Android client while exit-node is in use results in unusable network

### ISSUE THREAD

**Issue by bauersimon** [tailscale/tailscale#12318]
### What is the issue?

When I revoke ACL rights to use an exit node for an Android client **while the client is using the exit node**, then the client is left unable to use it's original network (i.e. I cannot `ping 1.1.1.1`, use WhatsApp, fetch GMail,... even though that was working fine).

I was able to reliably reproduce this unwanted state multiple times.

### Steps to reproduce

1. Verify internet access works on Android client (via [`termux`](https://termux.dev/): `ping 1.1.1.1` ✔️)
1. Allow the Android client exit node access in ACL settings by applying tag `vpn-client` in admin panel:
  ```json
  {
      "action": "accept",
      "src":    ["tag:vpn-client"],
      "dst":    ["autogroup:internet:*"],
  }
  ```
2. Select an exit node in Android client in the tailscale app
3. Verify internet access via exit node works on Android client (via [`termux`](https://termux.dev/): `ping 1.1.1.1` ✔️)
4. Revoke exit node ACL setting (basically "untag" Android client `vpn-client` in admin panel)
5. The tailscale app now shows "EXIT NODE: none"
6. Observe that "original" internet access is broken on Android client (via [`termux`](https://termux.dev/): `ping 1.1.1.1` ❌)
7. Re-grant exit node ACL setting in admin panel (tag `vpn-client` again)
8. The tailscale app now shows "EXIT NODE: `<whatever-previously-selected-exit-node>`"
9. Disable exit node in Android client in the tailscale app
10. Verify internet access works again on Android client (via [`termux`](https://termux.dev/): `ping 1.1.1.1` ✔️)

### OS

Android

### OS version

Android 12 (Samsung Galaxy Core UI 4.1)

### Tailscale version

1.66.3-teae73f821-g63a2bbc8e71

### Bug report

BUG-4b941a28b301993ba9f311e4584d7baa687f6c73ba354435bf25c697b9d232d6-20240601154122Z-6441b4659149902f

**Comment by agottardo** [comment]
This is an edge case that wasn't considered in the mobile apps UI. What do you think the expected behavior here should be after an admin removes your access to an exit node? Because I think there are two solutions to this:

1. We leave you unable to access the Internet as we are currently doing, but with a better error message ("the exit node previously selected is no longer available, please select another one to restore connectivity").
2. We detect that access was revoked and delete the exit node from settings (the same as tapping "None" in the UI). The consequence here is that if you were expecting your traffic to be routed through the exit node, connectivity will be restored, but suddenly all your traffic is going to get routed through your public IP.

Some customers might say that in the event an admin misconfigures their ACLs, revoking access to an exit node shouldn't be fair game for all the client traffic to be immediately routed through the public Internet. Essentially... you are asking for it to be fail-open instead of the current fail-closed behavior.

**Comment by bradfitz** [comment]
Yeah, this is working as intended for safety. But the UI could be improved to tell you what happened.

**Comment by bauersimon** [comment]
Thanks for the insight. It makes sense that the Android client does not stay connected to the exit node once access is revoked.

But I don't think this is solely a UI problem.

The problem is that the app then displays "exit node: none" but then stays in some **invalid state** where network access to the local LAN is entirely broken (see step `6.`). What is even worse: the only way to get back to a valid working LAN state that I found is to temporarily allow ACL access to the exit node again so it can be deselected from the app (see step `9.`). I guess one other solution would be to reinstall the app?

But the intended behavior I would expect is that the client just disconnects from the exit node and falls back to normal LAN connection. Or that there is a UI warning but then also a button to "disconnect from the invalid exit node". I.e. what I would expect is that in step `6.` above I can just `ping 1.1.1.1` via LAN just as I was able to in step `1.`.

I guess **if this is definitely intended behavior**. Then how do I reach step `10.` above without doing step `7.`?

**Comment by agottardo** [comment]
@bauersimon you've made the problem clear: the UX/UI around this is broken. And the Android app currently doesn't let you select None when it happens. It's an hard-to-hit edge case we didn't account for as I explained in my previous message. We're going to fix it by fixing the UI (fix the None issue, and show a warning with a "disconnect from the invalid exit node" button).

> the intended behavior I would expect is that the client just disconnects from the exit node and falls back to normal LAN connection

^^ This we're not going to do, as it introduces a vulnerability in the security model.

**Comment by bauersimon** [comment]
Alrightsy - cool! Sorry for being so overly explicit. The "this is intended"-part just got me worried that everything was goint to stay as it is. Thanks for taking the time to look into this.

**Comment by dblohm7** [comment]
@agottardo, assigning you just to keep this on your radar for future improvements to GUI health reporting
